storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -512,10 +514,146 @@ func (s *peersSource) ReadIndex(
+		return nil, err
+	}
+
+	var (
+		count         = len(shardsTimeRanges)
+		concurrency   = s.opts.DefaultShardConcurrency()
+		dataBlockSize = ns.Options().RetentionOptions().BlockSize()
+		resultOpts    = s.opts.ResultOptions()
+		idxOpts       = ns.Options().IndexOptions()
+		version       = s.opts.FetchBlocksMetadataEndpointVersion()
+		resultLock    = &sync.Mutex{}
+		wg            sync.WaitGroup
+	)
+	s.log.WithFields(
+		xlog.NewField("shards", count),
+		xlog.NewField("concurrency", concurrency),
+	).Infof("peers bootstrapper bootstrapping index for ranges")
+
+	workers := xsync.NewWorkerPool(concurrency)
+	workers.Init()
+	for shard, ranges := range shardsTimeRanges {
+		shard, ranges := shard, ranges
+		wg.Add(1)
+		workers.Go(func() {
+			defer wg.Done()
+
+			iter := ranges.Iter()
+			for iter.Next() {
+				currRange := iter.Value()
+				metadata, err := session.FetchBootstrapBlocksMetadataFromPeers(ns.ID(),
+					shard, currRange.Start, currRange.End, resultOpts, version)
+				if err != nil {
+					// Make this period unfulfilled
+					s.markIndexResultErrorAsUnfulfilled(r, resultLock, err,
+						shard, currRange)
+					continue
+				}
+
+				for metadata.Next() {
+					_, dataBlock := metadata.Current()
+					dataBlockRange := xtime.Range{
+						Start: dataBlock.Start,
+						End:   dataBlock.Start.Add(dataBlockSize),
+					}
+
+					inserted, err := s.readBlockMetadataAndIndex(r, resultLock, dataBlock,
+						idxOpts, resultOpts)
+					if err != nil {
+						// Make this period unfulfilled
+						s.markIndexResultErrorAsUnfulfilled(r, resultLock, err,
+							shard, dataBlockRange)
+					}
+
+					if !inserted {
+						// If the metadata wasn't inserted we finalize the metadata.
+						dataBlock.Finalize()
+					}
+				}
+				if err := metadata.Err(); err != nil {
+					// Make this period unfulfilled
+					s.markIndexResultErrorAsUnfulfilled(r, resultLock, err,
+						shard, currRange)
+				}
+			}
+		})
+	}
+
+	wg.Wait()
+
+	return r, nil
+}
+
+func (s *peersSource) readBlockMetadataAndIndex(
+	r result.IndexBootstrapResult,
+	resultLock *sync.Mutex,
+	dataBlock block.Metadata,
+	idxOpts namespace.IndexOptions,
+	resultOpts result.Options,
+) (bool, error) {
+	resultLock.Lock()
+	defer resultLock.Unlock()
+
+	segment, err := r.IndexResults().GetOrAddSegment(dataBlock.Start,
+		idxOpts, resultOpts)
+	if err != nil {
+		return false, err
+	}
+
+	exists, err := segment.ContainsID(dataBlock.ID.Bytes())
+	if err != nil {
+		return false, err
+	}
+	if exists {
+		return false, nil
+	}
+
+	d, err := convert.FromMetric(dataBlock.ID, dataBlock.Tags)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = segment.Insert(d)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (s *peersSource) markIndexResultErrorAsUnfulfilled(
+	r result.IndexBootstrapResult,
+	resultLock *sync.Mutex,
+	err error,
+	shard uint32,
+	timeRange xtime.Range,
+) {
+	// NB(r): We explicitly do not remove entries from the index results
```

Comment:
can you update this comment to talk about peer metadata instead of data files

Reply:
This is for peer metadata, thats why I mention "index results".  I can change it if that's not very clear though?

Reply:
> 	// It would also be difficult to remove only series that were added to the
 	// index block as results from data files can be subsets of the index block
 	// and there's no way to definitively delete the entry we added as a result
 	// of just this data file failing.

Reply:
"data files"

Reply:
I just meant that you use the word data files, which I thought was just copy-pasta from the other bootstrapper, but I guess technically the metadata from peers originates in fileset files in some cases.

Reply:
Fair, yeah I'll say "series that were added to the index block as results from data blocks received from peers".

It was definitely a copy paste heh.

Reply:
Updated.